### PR TITLE
Avoid using the same LocationFactory in each operation

### PR DESCRIPTION
**Pull request by anwar6953**
See the JIRA for more information.
In summary, this is to avoid reusing the same FileContext (which caches a UGI), when we are impersonating as other users.

https://issues.cask.co/browse/CDAP-6606
http://builds.cask.co/browse/CDAP-DUT4470-3

### CODE REVIEW

**Comment by anwar6953** [comment]
This is under the assumption that we no longer use HDFSLocation (it is deprecated).

Edit: verified that we don't.

File: cdap-common/src/main/java/co/cask/cdap/common/namespace/DefaultNamespacedLocationFactory.java [hunk 1]
```diff
@@ -96,11 +98,27 @@ public Location get(Id.Namespace namespaceId, @Nullable String subPath) throws I
...
+    return cloneIfNecessary(locationFactory).create(namespaceDir).append(namespaceId.getId());
   }
 
   @Override
   public Location getBaseLocation() throws IOException {
-    return locationFactory.create("/");
+    return cloneIfNecessary(locationFactory).create("/");
+  }
+
+  // Creates a clone, if necessary (if it is a or delegates to a FileContextLocationFactory)
```
Comment (bdmogal) [inline cdap-common/src/main/java/co/cask/cdap/common/namespace/DefaultNamespacedLocationFactory.java]: something's not right about this comment.

File: cdap-common/src/main/java/co/cask/cdap/common/namespace/DefaultNamespacedLocationFactory.java [hunk 2]
```diff
@@ -96,11 +98,27 @@ public Location get(Id.Namespace namespaceId, @Nullable String subPath) throws I
...
+  // note that if it the passed LocationFactory object is a ForwardingLocationFactory, the delegate (or a clone of it)
```
Comment (bdmogal) [inline cdap-common/src/main/java/co/cask/cdap/common/namespace/DefaultNamespacedLocationFactory.java]: better to just turn this into a javadoc?

**Comment by chtyim** [comment]
LGTM

**Comment by anwar6953** [comment]
Tested on cluster; pending build.

**Comment by sreevatsanraman** [comment]
Build is green. Merging.

**Pull request merged**